dev/bots/analyze_snippet_code.dart

```diff
@@ -431,346 +725,260 @@ class _SnippetChecker {
+      stdout.removeAt(0);
+    }
+    if (stdout.isNotEmpty && stdout.first.startsWith('Running "flutter pub get" in ')) {
+      stdout.removeAt(0);
+    }
+    return stdout;
   }
 }
 
 /// A class to represent a line of input code.
 class _Line {
-  const _Line({this.code = '', required this.filename, this.line = -1, this.indent = 0})
+  const _Line({this.code = '', this.line = -1, this.indent = 0})
       : generated = false;
-  const _Line.generated({this.code = '', required this.filename, this.line = -1, this.indent = 0})
-      : generated = true;
+  const _Line.generated({this.code = ''})
+      : line = -1,
+        indent = 0,
+        generated = true;
 
-  /// The file that this line came from, or the file that the line was generated for, if [generated] is true.
-  final String filename;
   final int line;
   final int indent;
   final String code;
   final bool generated;
 
-  String toStringWithColumn(int column) {
-    if (column != null && indent != null) {
-      return '$filename:$line:${column + indent}: $code';
-    }
-    return toString();
+  String asLocation(String filename, int column) {
+    return '$filename:$line:${column + indent}';
   }
 
   @override
-  String toString() => '$filename:$line: $code';
+  String toString() => code;
 }
 
-/// A class to represent a section of snippet code, marked by "{@tool snippet}...{@end-tool}".
-class _Section {
-  const _Section(this.code, {this.dartVersionOverride});
-  factory _Section.combine(List<_Section> sections) {
-    final List<_Line> code = sections
-        .expand((_Section section) => section.code)
-        .toList();
-    return _Section(code);
-  }
-  factory _Section.fromStrings(_Line firstLine, List<String> code) {
-    final List<_Line> codeLines = <_Line>[];
-    for (int i = 0; i < code.length; ++i) {
-      codeLines.add(
-        _Line(
-          code: code[i],
-          filename: firstLine.filename,
-          line: firstLine.line + i,
-          indent: firstLine.indent,
-        ),
-      );
+/// A class to represent a section of snippet code, marked by "```dart ... ```", that ends up
+/// in a file we then analyze (each snippet is in its own file).
+class _SnippetFile {
+  const _SnippetFile(this.code, this.generatorComment, this.filename, this.indexLine);
+
+  factory _SnippetFile.fromLines(
+    List<_Line> code,
+    List<_Line> headers,
+    List<_Line> preamble,
+    String generatorComment,
+    String filename,
+  ) {
+    while (code.isNotEmpty && code.last.code.isEmpty) {
+      code.removeLast();
     }
-    return _Section(codeLines);
+    assert(code.isNotEmpty);
+    final _Line firstLine = code.firstWhere((_Line line) => !line.generated);
+    return _SnippetFile(
+      <_Line>[
+        ...headers,
+        const _Line.generated(), // blank line
+        if (preamble.isNotEmpty)
+          ...preamble,
+        if (preamble.isNotEmpty)
+          const _Line.generated(), // blank line
```

Comment:
```suggestion
        if (preamble.isNotEmpty) ...[
          ...preamble,
          const _Line.generated(), // blank line
        ],
```

Reply:
i considered doing that but isn't that less efficient? now you're building a list just to inject it back into another list, rather than just evaluating a boolean variable twice.